File: js/src/agents2.ts
import fetch from "cross-fetch";

const DEFAULT_BASE_URL = "http://localhost:3005";

export type ClientConfig = {
    apiKey: string;
    baseUrl?: string;
    timeoutMs?: number;
};

export type ApiResponse<T = any> = {
    statusCode: number;
    headers: Record<string, string>;
    data: T;
};

function buildUrl(baseUrl: string, path: string): string {
    const url = new URL(path.startsWith("/") ? path : `/${path}`, baseUrl.endsWith("/") ? baseUrl : `${baseUrl}/`);
    return url.toString();
}

function headersToObject(headers: any): Record<string, string> {
    const out: Record<string, string> = {};
    headers.forEach((v: string, k: string) => {
        out[k] = v;
    });
    return out;
}

export type Agents2CreateRequest = Record<string, any>;
export type Agents2UpdateRequest = Record<string, any>;

export async function createAgent2(
    { apiKey, baseUrl = DEFAULT_BASE_URL, timeoutMs }: ClientConfig,
    organizationId: string,
    body: Agents2CreateRequest
): Promise<ApiResponse> {
    const xOrg = organizationId || (typeof process !== "undefined" ? (process as any).env?.NEXA_ORGANIZATION_ID : undefined);
    if (!xOrg) throw new Error("organizationId is required (set NEXA_ORGANIZATION_ID or pass organizationId)");
    if (!body || typeof body !== "object") throw new Error("body is required");

    const controller = typeof AbortController !== "undefined" && timeoutMs ? new AbortController() : undefined;
    const timer = controller && timeoutMs ? setTimeout(() => controller.abort(), timeoutMs) : undefined;
    try {
        const response = await fetch(buildUrl(baseUrl, "/agents2"), {
            method: "POST",
            headers: {
                Authorization: `Bearer ${apiKey}`,
                Accept: "application/json",
                "Content-Type": "application/json",
                "Organization-Id": xOrg,
                "X-Organization-Id": xOrg,
                "x-organization-id": xOrg,
            },
            body: JSON.stringify(body),
            signal: controller?.signal,
        });
        const text = await response.text();
        let data: any;
        try { data = text ? JSON.parse(text) : null; } catch { data = text; }
        if (!response.ok) {
            throw new Error(`Agents2 create failed (${response.status}): ${text}`);
        }
        return { statusCode: response.status, headers: headersToObject(response.headers), data };
    } finally {
        if (timer) clearTimeout(timer);
    }
}

export async function listAgents2(
    { apiKey, baseUrl = DEFAULT_BASE_URL, timeoutMs }: ClientConfig,
    organizationId?: string,
): Promise<ApiResponse> {
    const xOrg = organizationId || (typeof process !== "undefined" ? (process as any).env?.NEXA_ORGANIZATION_ID : undefined);
    if (!xOrg) throw new Error("organizationId is required (set NEXA_ORGANIZATION_ID or pass organizationId)");

    const controller = typeof AbortController !== "undefined" && timeoutMs ? new AbortController() : undefined;
    const timer = controller && timeoutMs ? setTimeout(() => controller.abort(), timeoutMs) : undefined;
    try {
        const response = await fetch(buildUrl(baseUrl, "/agents2"), {
            headers: {
                Authorization: `Bearer ${apiKey}`,
                Accept: "application/json",
                "Organization-Id": xOrg,
                "X-Organization-Id": xOrg,
                "x-organization-id": xOrg,
            },
            signal: controller?.signal,
        });
        const text = await response.text();
        let data: any;
        try { data = text ? JSON.parse(text) : null; } catch { data = text; }
        if (!response.ok) {
            throw new Error(`Agents2 list failed (${response.status}): ${text}`);
        }
        return { statusCode: response.status, headers: headersToObject(response.headers), data };
    } finally {
        if (timer) clearTimeout(timer);
    }
}

export async function getAgent2ById(
    { apiKey, baseUrl = DEFAULT_BASE_URL, timeoutMs }: ClientConfig,
    id: string,
    organizationId?: string,
): Promise<ApiResponse> {
    if (!id) throw new Error("id is required");
    const xOrg = organizationId || (typeof process !== "undefined" ? (process as any).env?.NEXA_ORGANIZATION_ID : undefined);
    if (!xOrg) throw new Error("organizationId is required (set NEXA_ORGANIZATION_ID or pass organizationId)");

    const controller = typeof AbortController !== "undefined" && timeoutMs ? new AbortController() : undefined;
    const timer = controller && timeoutMs ? setTimeout(() => controller.abort(), timeoutMs) : undefined;
    try {
        const response = await fetch(buildUrl(baseUrl, `/agents2/${encodeURIComponent(id)}`), {
            headers: {
                Authorization: `Bearer ${apiKey}`,
                Accept: "application/json",
                "Organization-Id": xOrg,
                "X-Organization-Id": xOrg,
                "x-organization-id": xOrg,
            },
            signal: controller?.signal,
        });
        const text = await response.text();
        let data: any;
        try { data = text ? JSON.parse(text) : null; } catch { data = text; }
        if (!response.ok) {
            throw new Error(`Agents2 get failed (${response.status}): ${text}`);
        }
        return { statusCode: response.status, headers: headersToObject(response.headers), data };
    } finally {
        if (timer) clearTimeout(timer);
    }
}

export async function updateAgent2(
    { apiKey, baseUrl = DEFAULT_BASE_URL, timeoutMs }: ClientConfig,
    id: string,
    organizationId: string,
    body: Agents2UpdateRequest
): Promise<ApiResponse> {
    if (!id) throw new Error("id is required");
    const xOrg = organizationId || (typeof process !== "undefined" ? (process as any).env?.NEXA_ORGANIZATION_ID : undefined);
    if (!xOrg) throw new Error("organizationId is required (set NEXA_ORGANIZATION_ID or pass organizationId)");
    if (!body || typeof body !== "object") throw new Error("body is required");

    const controller = typeof AbortController !== "undefined" && timeoutMs ? new AbortController() : undefined;
    const timer = controller && timeoutMs ? setTimeout(() => controller.abort(), timeoutMs) : undefined;
    try {
        const response = await fetch(buildUrl(baseUrl, `/agents2/${encodeURIComponent(id)}`), {
            method: "PATCH",
            headers: {
                Authorization: `Bearer ${apiKey}`,
                Accept: "application/json",
                "Content-Type": "application/json",
                "Organization-Id": xOrg,
                "X-Organization-Id": xOrg,
                "x-organization-id": xOrg,
            },
            body: JSON.stringify(body),
            signal: controller?.signal,
        });
        const text = await response.text();
        let data: any;
        try { data = text ? JSON.parse(text) : null; } catch { data = text; }
        if (!response.ok) {
            throw new Error(`Agents2 update failed (${response.status}): ${text}`);
        }
        return { statusCode: response.status, headers: headersToObject(response.headers), data };
    } finally {
        if (timer) clearTimeout(timer);
    }
}

File: python/sdk/agents2.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Any, Dict, Optional
import os
import requests


@dataclass
class ApiResponse:
    status_code: int
    headers: Dict[str, str]
    data: Any


class Agents2Client:
    BASE_URL = "http://localhost:3005"

    def __init__(self, api_key: str, base_url: Optional[str] = None, timeout: Optional[float] = 30.0,
                 default_organization_id: Optional[str] = None):
        if not api_key:
            raise ValueError("API key is required")
        self.api_key = api_key
        self.base_url = (base_url or self.BASE_URL).rstrip("/")
        self.timeout = timeout
        self.default_org_id = default_organization_id or os.getenv("NEXA_ORGANIZATION_ID")
        self._base_headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Accept": "application/json",
        }

    def _resolve_org(self, organization_id: Optional[str]) -> str:
        org = organization_id or self.default_org_id or os.getenv("NEXA_ORGANIZATION_ID")
        if not org:
            raise ValueError("organization_id is required")
        return org

    def _headers(self, organization_id: str, content_type: Optional[str] = None) -> Dict[str, str]:
        headers = dict(self._base_headers)
        if content_type:
            headers["Content-Type"] = content_type
        # Set multiple casings for maximum compatibility
        headers["Organization-Id"] = organization_id
        headers["X-Organization-Id"] = organization_id
        headers["x-organization-id"] = organization_id
        return headers

    def create_agent(self, *, organization_id: Optional[str], body: Dict[str, Any]) -> ApiResponse:
        org = self._resolve_org(organization_id)
        if not isinstance(body, dict):
            raise ValueError("body must be a JSON object (dict)")
        resp = requests.post(
            f"{self.base_url}/agents2",
            headers=self._headers(org, content_type="application/json"),
            json=body,
            timeout=self.timeout,
        )
        try:
            data = resp.json()
        except Exception:
            data = resp.text
        if not resp.ok:
            raise RuntimeError(f"Agents2 create failed ({resp.status_code}): {data}")
        return ApiResponse(status_code=resp.status_code, headers=dict(resp.headers), data=data)

    def list_agents(self, organization_id: Optional[str] = None) -> ApiResponse:
        org = self._resolve_org(organization_id)
        resp = requests.get(
            f"{self.base_url}/agents2",
            headers=self._headers(org),
            timeout=self.timeout,
        )
        try:
            data = resp.json()
        except Exception:
            data = resp.text
        if not resp.ok:
            raise RuntimeError(f"Agents2 list failed ({resp.status_code}): {data}")
        return ApiResponse(status_code=resp.status_code, headers=dict(resp.headers), data=data)

    def get_agent_by_id(self, id: str, organization_id: Optional[str] = None) -> ApiResponse:
        if not id:
            raise ValueError("id is required")
        org = self._resolve_org(organization_id)
        resp = requests.get(
            f"{self.base_url}/agents2/{id}",
            headers=self._headers(org),
            timeout=self.timeout,
        )
        try:
            data = resp.json()
        except Exception:
            data = resp.text
        if not resp.ok:
            raise RuntimeError(f"Agents2 get failed ({resp.status_code}): {data}")
        return ApiResponse(status_code=resp.status_code, headers=dict(resp.headers), data=data)

    def update_agent(self, id: str, *, organization_id: Optional[str], body: Dict[str, Any]) -> ApiResponse:
        if not id:
            raise ValueError("id is required")
        org = self._resolve_org(organization_id)
        if not isinstance(body, dict):
            raise ValueError("body must be a JSON object (dict)")
        resp = requests.patch(
            f"{self.base_url}/agents2/{id}",
            headers=self._headers(org, content_type="application/json"),
            json=body,
            timeout=self.timeout,
        )
        try:
            data = resp.json()
        except Exception:
            data = resp.text
        if not resp.ok:
            raise RuntimeError(f"Agents2 update failed ({resp.status_code}): {data}")
        return ApiResponse(status_code=resp.status_code, headers=dict(resp.headers), data=data)

